dbms/src/Storages/KVStore/ProxyStateMachine.h

```diff
@@ -0,0 +1,421 @@
+    /// Then, inform proxy to stop by setting `tiflash_instance_wrap.status`.
+    void stopProxy(TMTContext & tmt_context)
+    {
+        if (!proxy_conf.is_proxy_runnable)
+        {
+            tmt_context.setStatusTerminated();
+            return;
+        }
+        if (proxy_conf.is_proxy_runnable && tiflash_instance_wrap.status != EngineStoreServerStatus::Running)
+        {
+            LOG_ERROR(log, "Current status of engine-store is NOT Running, should not happen");
+            exit(-1);
+        }
+        LOG_INFO(log, "Set store context status Stopping");
+        tmt_context.setStatusStopping();
+        {
+            // Wait until there is no read-index task.
+            while (tmt_context.getKVStore()->getReadIndexEvent())
+                std::this_thread::sleep_for(std::chrono::milliseconds(200));
+        }
+        tmt_context.setStatusTerminated();
+        tmt_context.getKVStore()->stopReadIndexWorkers();
+        LOG_INFO(log, "Set store context status Terminated");
+        {
+            // update status and let proxy stop all services except encryption.
+            tiflash_instance_wrap.status = EngineStoreServerStatus::Stopping;
+            LOG_INFO(log, "Set engine store server status Stopping");
+        }
+        // wait proxy to stop services
+        if (proxy_conf.is_proxy_runnable)
+        {
+            LOG_INFO(log, "Let tiflash proxy to stop all services");
+            while (tiflash_instance_wrap.proxy_helper->getProxyStatus() != RaftProxyStatus::Stopped)
+                std::this_thread::sleep_for(std::chrono::milliseconds(200));
+            LOG_INFO(log, "All services in tiflash proxy are stopped");
+        }
+    }
+
+    // TMTContext can not be accessed now.
+    void destroyProxyContext()
+    {
+        if (!proxy_conf.is_proxy_runnable)
+            return;
+
+        LOG_INFO(log, "Unlink tiflash_instance_wrap.tmt");
+        // Reset the `tiflash_instance_wrap.tmt` before `global_context` get released, or it will be a dangling pointer
+        tiflash_instance_wrap.tmt = nullptr;
+    }
+
+    /// Inform proxy to shutdown, and join the thread.
+    void waitProxyStopped()
+    {
+        if (!proxy_conf.is_proxy_runnable)
+            return;
+
+        LOG_INFO(log, "Let tiflash proxy shutdown");
+        tiflash_instance_wrap.status = EngineStoreServerStatus::Terminated;
+        tiflash_instance_wrap.tmt = nullptr;
+        LOG_INFO(log, "Wait for tiflash proxy thread to join");
+        proxy_runner->join();
+        LOG_INFO(log, "tiflash proxy thread is joined");
+    }
+
+    ProxyStateMachine(LoggerPtr log_, TiFlashProxyConfig && proxy_conf_)
+        : log(std::move(log_))
+        , proxy_conf(std::move(proxy_conf_))
+    {
+        helper = GetEngineStoreServerHelper(&tiflash_instance_wrap);
+        proxy_runner = std::make_unique<RaftStoreProxyRunner>(
+            RaftStoreProxyRunner::RunRaftStoreProxyParms{&helper, proxy_conf},
+            log);
+    }
```

Comment:
nit. IMO, make the constructor function being the first function of this class is better for code reading

Reply:
Move forward constructors of TiFlashProxyConfig and ProxyStateMachine